libraries/griptape_nodes_library/griptape_nodes_library/number/math_expression.py

```
@@ -0,0 +1,399 @@
+        )
+        expression = re.sub(
+            r"\)(" + number_pattern + r"|" + var_pattern + r")(?=\s|$|[^a-zA-Z0-9.])", r")*\1", expression
+        )
+
+        # 4. Numbers before variables: 2a -> 2*a, 2.5a -> 2.5*a
+        expression = re.sub(
+            r"(" + number_pattern + r")\s*(" + var_pattern + r")(?=\s|$|[^a-zA-Z0-9.])", r"\1*\2", expression
+        )
+        expression = re.sub(
+            r"(" + number_pattern + r")(" + var_pattern + r")(?=\s|$|[^a-zA-Z0-9.])", r"\1*\2", expression
+        )
+
+        # 5. Variables before numbers: a2 -> a*2, a2.5 -> a*2.5
+        expression = re.sub(
+            r"\b(" + var_pattern + r")\s*(" + number_pattern + r")(?=\s|$|[^a-zA-Z0-9.])", r"\1*\2", expression
+        )
+        expression = re.sub(
+            r"\b(" + var_pattern + r")(" + number_pattern + r")(?=\s|$|[^a-zA-Z0-9.])", r"\1*\2", expression
+        )
+
+        # 6. Adjacent variables: ab -> a*b, abc -> a*b*c (iterative to handle chains)
+        for _ in range(10):  # Max 10 iterations should handle long chains
+            new_expr = re.sub(
+                r"\b(" + var_pattern + r")\s*(" + var_pattern + r")(?=\s|$|[^a-zA-Z])", r"\1*\2", expression
+            )
+            new_expr = re.sub(r"\b(" + var_pattern + r")(" + var_pattern + r")(?=\s|$|[^a-zA-Z])", r"\1*\2", new_expr)
+            if new_expr == expression:
+                break
+            expression = new_expr
+
+        # 7. Clean up: remove extra asterisks and spaces
+        expression = re.sub(r"\*\*\*+", r"*", expression)  # *** -> *
+        expression = re.sub(r"(?<!\*)\s*\*\s*(?!\*)", r"*", expression)  # Clean spaces around *
+
+        return expression
+
+    def _evaluate_expression(self, expression: str) -> float | int:  # noqa: PLR0911
+        """Safely evaluate the mathematical expression using asteval."""
+        if not expression or not expression.strip():
+            return 0.0
```

Is 0 representing an error be clear to the user? Can we instead just surface these errors?

---

this isn't necessarily an error though - it's if they don't include an expression or it's blank, right?

so it just returns 0?